harness/tests/experiment/pytorch_utils.py

```diff
@@ -0,0 +1,265 @@
+    core_context.preempt.start()
+    if core_context._tensorboard_manager is not None:
+        core_context._tensorboard_manager.start()
+
+    core_context.train._trial_id = "1"
+    distributed_backend = det._DistributedBackend()
+    if expose_gpus:
+        gpu_uuids = gpu.get_gpu_uuids()
+    else:
+        gpu_uuids = []
+
+    pytorch._PyTorchTrialController.pre_execute_hook(trial_seed, distributed_backend)
+    trial_context = pytorch.PyTorchTrialContext(
+        core_context=core_context,
+        trial_seed=trial_seed,
+        hparams=hparams,
+        slots_per_trial=slots_per_trial,
+        num_gpus=len(gpu_uuids),
+        exp_conf=exp_config,
+        aggregation_frequency=aggregation_frequency,
+        steps_completed=steps_completed,
+        managed_training=True,  # this must be True to put model on GPU
+        debug_enabled=False,
+    )
+    trial_context._set_default_gradient_compression(False)
+    trial_context._set_default_average_aggregated_gradients(True)
+    trial_inst = trial_class(trial_context)
+
+    trial_controller = pytorch._PyTorchTrialController(
+        trial_inst=trial_inst,
+        context=trial_context,
+        max_length=pytorch.Batch(max_batches),
+        checkpoint_period=pytorch.Batch(min_checkpoint_batches),
+        validation_period=pytorch.Batch(min_validation_batches),
+        searcher_metric_name=trial_class._searcher_metric,
+        reporting_period=pytorch.Batch(scheduling_unit),
+        local_training=True,
+        latest_checkpoint=latest_checkpoint,
+        steps_completed=steps_completed,
+        smaller_is_better=bool(exp_config["searcher"]["smaller_is_better"]),
+        test_mode=False,
+        checkpoint_policy=exp_config["checkpoint_policy"],
+        step_zero_validation=bool(exp_config["perform_initial_validation"]),
+        det_profiler=None,
+        global_batch_size=None,
+    )
+
+    trial_controller._set_data_loaders()
+    trial_controller.state = pytorch._TrialState()
+
+    trial_controller.training_iterator = iter(trial_controller.training_loader)
+    return trial_inst, trial_controller
+
+
+def train_for_checkpoint(
+    hparams: typing.Dict,
+    trial_class: pytorch.PyTorchTrial,
+    tmp_path: pathlib.Path,
+    exp_config: typing.Dict,
+    slots_per_trial: int = 1,
+    steps: int = 1,
+) -> int:
+    checkpoint_dir = str(tmp_path.joinpath("checkpoint"))
+    tensorboard_path = tmp_path.joinpath("tensorboard")
+
+    trial, trial_controller = create_trial_and_trial_controller(
+        trial_class=trial_class,
+        hparams=hparams,
+        slots_per_trial=slots_per_trial,
+        exp_config=exp_config,
+        max_batches=steps,
+        min_validation_batches=steps,
+        min_checkpoint_batches=steps,
+        checkpoint_dir=checkpoint_dir,
+        tensorboard_path=tensorboard_path,
+        expose_gpus=True,
+    )
+
+    trial_controller.run()
+
+    assert len(os.listdir(checkpoint_dir)) == 1, "trial did not create a checkpoint"

```

Comment:
since you've broken up the functions, perhaps assert that this function adds one checkpoint to checkpoint_dir, rather than asserting that there be exactly one?

Same for train_from_checkpoint.

Reply:
I think this is fine for `train_for_checkpoint`, it'll always create 1 and the directory doesn't necessarily even exist before the trial runs. However I did update this in `train_for_checkpoint`, so that it'll be reusable if necessary. Good catch.